### PR TITLE
Add an unrecoverable exception that will move messages to the error queue bypassing retries

### PR DESCRIPTION
**Pull request by lailabougria**
Implements https://github.com/Particular/NServiceBus/issues/6421

### CODE REVIEW

**Review by timbussmann — COMMENTED** [review]
a few nitpicks

File: src/NServiceBus.Core/Recoverability/UnrecoverableException.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Runtime.Serialization;
+
+    /// <summary>
+    /// This exception will bypass any retries and will cause the message to be moved to the error queue immediately.
+    /// </summary>
+    [Serializable]
+    public class UnrecoverableException : Exception
```
[the remark this answers is not in the document]
Comment (lailabougria) [reply src/NServiceBus.Core/Recoverability/UnrecoverableException.cs:10]: Let's keep it for now, and we can see if any other changes make us change our mind? For now, I can't think of a better name to be honest. Unless maybe UnretryableException, but that sounds off.

File: src/NServiceBus.AcceptanceTests/Core/Recoverability/When_unrecoverable_exception_thrown.cs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    public class When_unrecoverable_exception_thrown : NServiceBusAcceptanceTest
+    {
+        static string ErrorSpyAddress => Conventions.EndpointNamingConvention(typeof(ErrorSpy));
+
+        [Test]
+        public async Task Should_move_to_error_queue_without_retries()
+        {
+            var context = await Scenario.Define<Context>()
+                .WithEndpoint<EndpointWithOutgoingMessages>(b => b.DoNotFailOnErrorMessages()
```
[the remark this answers is not in the document]
Comment (lailabougria) [reply src/NServiceBus.AcceptanceTests/Core/Recoverability/When_unrecoverable_exception_thrown.cs:18]: This hurts my eyes. Inlining is so personal...

File: src/NServiceBus.AcceptanceTests/Core/Recoverability/When_unrecoverable_exception_thrown.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+            var failedMessages = context.FailedMessages.Single().Value;
+            Assert.AreEqual(1, failedMessages.Count);
+            Assert.IsInstanceOf<SpecificException>(failedMessages.First().Exception);
+            Assert.AreEqual(1, context.NrOfTimesHandlerWasInvoked);
+        }
+
+        class Context : ScenarioContext
+        {
+            public bool MessageMovedToErrorQueue { get; set; }
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Core/Recoverability/When_unrecoverable_exception_thrown.cs:47]: no longer needed I think?

```suggestion
```

**Review by timbussmann — APPROVED** [review]

**Comment by lailabougria** [comment]
Closing this as further analysis has shown that this might not be the right way to evolve the recoverability API.

**Pull request closed without merging**